core/src/main/java/io/grpc/LoadBalancer2.java

```diff
@@ -0,0 +1,327 @@
+
+    /**
+     * Out-of-band channel for LoadBalancer’s own RPC needs, e.g., talking to an external
+     * load-balancer service.
+     *
+     * <p>The LoadBalancer is responsible for closing unused OOB channels, and closing all OOB
+     * channels within {@link #shutdown}.
+     */
+    public abstract ManagedChannel createOobChannel(
+        EquivalentAddressGroup eag, String authority);
+
+    /**
+     * Set a new picker to the channel.
+     *
+     * <p>When a new picker is provided via {@link Helper#updatePicker}, the channel will apply the
+     * picker on all buffered RPCs, by calling {@link SubchannelPicker#pickSubchannel}.
+     *
+     * <p>The channel will hold the a picker and use it for all RPCs, until {@link #updatePicker} is
+     * called again and a new picker replaces the old one.  If {@link #updatePicker} has never been
+     * called, the channel will buffer all RPCs until a picker is provided.
+     */
+    public abstract void updatePicker(SubchannelPicker picker);
+
+    /**
+     * Schedule a task to be run in the Channel Executor, which serializes the task with the
+     * callback methods on the {@link LoadBalancer2} interface.
+     */
+    public abstract void runSeralized(Runnable task);
+
+    /**
+     * Returns the NameResolver of the channel.
+     */
+    public abstract NameResolver.Factory getNameResolverFactory();
+
+    /**
+     * Returns the authority string of the channel, which is derived from the DNS-style target name.
+     */
+    public abstract String getAuthority();
+  }
+
+  /**
+   * A logical connection to a server, or a group of equivalent servers represented by an {@link
+   * EquivalentAddressGroup}.
+   *
+   * <p>It maintains at most one physical connection (aka transport) for sending new RPCs, while
+   * also keeps track of previous transports that has been shut down but not terminated yet.
+   *
+   * <p>If there isn't an active transport yet, and the Subchannel is assigned to an RPC, it will
+   * create a new transport.  It won't actively create transports otherwise.  {@link
+   * #requestConnection} can be used to ask Subchannel to create a transport if there isn't any.
+   */
+  @ThreadSafe
+  public abstract static class Subchannel {
+    /**
+     * Shuts down the Subchannel.  No new RPCs will be accepted.
+     */
+    public abstract void shutdown();
+
+    /**
+     * Asks the Subchannel to create a connection (aka transport), if there isn't an active one.
+     */
+    public abstract void requestConnection();
+
+    /**
+     * Returns the addresses that this Subchannel is bound to.
+     */
+    public abstract EquivalentAddressGroup getAddresses();
+
+    /**
+     * The same attributes passed to {@link #createSubChannel}. LoadBalancer can use it to attach
+     * additional information here, e.g., the shard this Subchannel belongs to.
+     */
+    public abstract Attributes getAttributes();
+  }
```

Comment:
`Attributes` are immutable which means that if we want to be able to update attributes on existing channel, we'll need `setAttributes(Attributes)`